### PR TITLE
Improved branch restructuring

### PR DESCRIPTION
**Pull request by ezrosent**
This PR adapts the algorithm from the original RVSDG papers for restructuring branches. This will let us translate more programs successfully.

I suspect there is more work to do on the translation back to CFGs now that this is up.

This PR also includes a few cleanups around making LLVM tests optional, and adding some switches to dump CFG visualizations to `/tmp`; both of those use cargo features.

### CODE REVIEW

File: src/rvsdg/restructure.rs [hunk 1]
```diff
@@ -306,93 +276,242 @@ impl SwitchCfgFunction {
...
 }
 
 const JMP: Branch = Branch {
     op: BranchOp::Jmp,
     pos: None,
 };
+
+#[derive(Default)]
+struct Continuation {
+    // Nodes in the "tail" that are targetted by an edge out of the given branch node.
```
[the remark this answers is not in the document]
Comment (oflatt) [reply src/rvsdg/restructure.rs:507]: I think the tail is the dominance frontier or something like that?

File: src/rvsdg/restructure.rs [hunk 2]
```diff
@@ -306,93 +276,242 @@ impl SwitchCfgFunction {
...
+        if cont.reentry_nodes.is_empty() {
+            // Nothing to do.
+            return;
+        }
+        // assert!(
+        //     !cont.reentry_nodes.is_empty(),
+        //     "No reentry nodes found [start: {start:?}, entry: {entry:?}, exit: {exit:?}]"
+        // );
+
+        if cont.reentry_nodes.len() == 1 {
```
Comment (oflatt) [inline src/rvsdg/restructure.rs:380]: Which of these if statements are "optimizations" that generate better cfgs and which ones are necessary? For example, is this one necessary or does the else case technically work?

File: src/rvsdg/restructure.rs [hunk 3]
```diff
@@ -306,93 +276,242 @@ impl SwitchCfgFunction {
...
+                    self.split_arc(edge);
+                    continue;
                 }
-            }
 
-            // The rest of this loop is focused on mux/demux for multiple
-            // continuations. If we only have one, we are done.
-            if conts.len() == 1 {
-                continue;
+                if reentry_edges.len() == 1 {
```
Comment (oflatt) [inline src/rvsdg/restructure.rs:403]: is this one an "optimization"?

File: src/rvsdg/restructure.rs [hunk 4]
```diff
@@ -306,93 +276,242 @@ impl SwitchCfgFunction {
...
+                    continue;
+                }
+                // There are multiple exit edges to the tail. Have them all join
+                // into an intermediate block.
+                let inter = self.fresh_block();
+                self.graph.add_edge(inter, tail, JMP);
+                for e in reentry_edges {
+                    self.divert_edge(*e, inter);
+                }
+                self.restructure_branches_inner(target, inter, state);
```
Comment (oflatt) [inline src/rvsdg/restructure.rs:421]: In this recursive case, do we necessarily hit the if statement that came before this?
If so, can that be a helper that we call here?

File: src/rvsdg/from_cfg.rs [hunk 5]
```diff
@@ -331,32 +399,14 @@ impl<'a> RvsdgBuilder<'a> {
...
-                        };
-                        !dom.any(|n| n == curr)
-                    })
-                    .nth(1)
-                    .is_some()
-                {
-                    break;
-                }
+            while curr != join_point {
                 curr = self.try_loop(curr)?.unwrap();
```
Comment (oflatt) [inline src/rvsdg/from_cfg.rs:410]: What does it mean for try_loop or try_branch to return None?

**Review by oflatt — CHANGES_REQUESTED** [review]
Some questions- I think I'll need to digest this algorithm better sometime, but it should't block the PR.
However, I'd love to see at least a few more tests before we merge. Perhaps more unstructured bril programs that don't involve pointers? In fact, we could randomly generate some cfgs and see what happens

File: src/rvsdg/restructure.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (ezrosent) [reply src/rvsdg/restructure.rs]: Added a comment here.

File: src/rvsdg/restructure.rs [hunk 6]
```diff
@@ -306,93 +276,242 @@ impl SwitchCfgFunction {
...
+    }
 }
 
 const JMP: Branch = Branch {
     op: BranchOp::Jmp,
     pos: None,
 };
+
+#[derive(Default)]
+struct Continuation {
```
[the remark this answers is not in the document]
Comment (ezrosent) [reply src/rvsdg/restructure.rs:542]: Done!

File: src/rvsdg/restructure.rs [hunk 7]
```diff
@@ -306,93 +276,242 @@ impl SwitchCfgFunction {
...
+                break;
+            };
+            if neighbors.next().is_some() {
+                break;
+            }
+            cur = next;
+        }
+        cur
+    }
+    fn divert_edge(&mut self, edge: EdgeIndex, new_target: NodeIndex) -> EdgeIndex {
```
[the remark this answers is not in the document]
Comment (ezrosent) [reply src/rvsdg/restructure.rs]: Done. I also renamed the method to `retarget`.

File: src/rvsdg/restructure.rs (hunk 3 above)
Comment (ezrosent) [reply src/rvsdg/restructure.rs:409]: Yup. Same answer as the other question.

File: src/rvsdg/restructure.rs (hunk 2 above)
Comment (ezrosent) [reply src/rvsdg/restructure.rs:386]: Added some comments here. This structure is also in the JLM repo: it avoids us adding extra nodes and extra predicates when we don't need to.

File: src/rvsdg/restructure.rs (hunk 4 above)
Comment (ezrosent) [reply src/rvsdg/restructure.rs:427]: I don't think so. We don't actually know anything about target here, just that it isn't dominated by any of the outgoing edges. Unless I'm misunderstanding which `if` you're talking about.

File: src/rvsdg/from_cfg.rs (hunk 5 above)
Comment (ezrosent) [reply src/rvsdg/from_cfg.rs:410]: It happens if you reach the exit block early.

**Comment by ezrosent** [comment]
> However, I'd love to see at least a few more tests before we merge. Perhaps more unstructured bril programs that don't involve pointers? In fact, we could randomly generate some cfgs and see what happens

There are 3 or 4 more failing CFGs in the issues that I'm hoping to fix in follow-ups, but I can look at adding some more.

**Comment by ezrosent** [comment]
I'm not sure where the failing test is from either... that snapshot is present on this branch.

**Review by oflatt — APPROVED** [review]
Great work!

**Pull request merged**